mllib/src/main/scala/org/apache/spark/mllib/util/MLUtils.scala

```diff
@@ -82,6 +82,19 @@ object MLUtils {
           val value = indexAndValue(1).toDouble
           (index, value)
         }.unzip
+
+        // check if indices is one-based and in ascending order
```

Comment:
indices(0) is compared with -1 in current implementation. 
Your way looks more compact. The only concern is if it will create many groupIterators. I'll run some benchmark and get back.